File: Matlab/Divided_Difference.m
function [div_diff] = Divided_Difference(func, X)
    %%% ---------------------------------------------------

    % Parameters
    % ----------
    % X       : Leja points
    % func    : func(X)

    % Returns
    % -------
    % div_diff : Polynomial coefficients

    %%% ---------------------------------------------------

    N = length(X) - 1;
    div_diff = func(X);

    for jj = 1 : N
        for ii = N + 1 : -1 : jj + 1
            div_diff(ii) = (div_diff(ii) - div_diff(ii - 1))/(X(ii) - X(ii - jj)); 
        end
    end

end

File: Divided_Difference.py
import numpy as np

def Divided_Difference(X, diffs):
    """
    Parameters
    ----------
    X       : Leja points
    diffs   : Phi function array

    Returns
    -------
    div_diff : Polynomial coefficients

    """

    N = len(X)
    div_diff = diffs
    
    for ii in range(1, N):
        div_diff[ii:N] = (div_diff[ii:N] - div_diff[ii - 1])/(X[ii:N] - X[ii - 1])

    return div_diff
